codec/types.go

```diff
@@ -0,0 +1,248 @@
+package codec
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gogo/protobuf/proto"
+)
+
+// Marshaler defines the interface module codecs must implement in order to support
+// backwards compatibility with Amino while allowing custom serialization. The
+// implementing type can extend this with ProtoMarshaler to support and utilize
+// protocol buffers.
+type Marshaler interface {
+	MarshalBinaryBare(o interface{}) ([]byte, error)
+	MustMarshalBinaryBare(o interface{}) []byte
+
+	MarshalBinaryLengthPrefixed(o interface{}) ([]byte, error)
+	MustMarshalBinaryLengthPrefixed(o interface{}) []byte
+
+	UnmarshalBinaryBare(bz []byte, ptr interface{}) error
+	MustUnmarshalBinaryBare(bz []byte, ptr interface{})
+
+	UnmarshalBinaryLengthPrefixed(bz []byte, ptr interface{}) error
+	MustUnmarshalBinaryLengthPrefixed(bz []byte, ptr interface{})
+
+	MarshalJSON(o interface{}) ([]byte, error) // nolint: stdmethods
+	MustMarshalJSON(o interface{}) []byte
+
+	UnmarshalJSON(bz []byte, ptr interface{}) error // nolint: stdmethods
+	MustUnmarshalJSON(bz []byte, ptr interface{})
+}
+
+// ProtoMarshaler defines an interface a type must implement as protocol buffer
+// defined message.
+type ProtoMarshaler interface {
+	proto.Message // for JSON serialization
+
+	Marshal() ([]byte, error)
+	MarshalTo(data []byte) (n int, err error)
+	MarshalToSizedBuffer(dAtA []byte) (int, error)
+	Size() int
+	Unmarshal(data []byte) error
+}
+
+// BaseCodec defines a codec implementing the Marshaler interface that allows
+// for Protocol buffer and Amino-based serialization. Module-level codecs should
+// use and extend this codec as necessary. If a module-level codec does not have
+// a need for rich interfaces, then they may simply use a BaseCodec as-is.
+type BaseCodec struct {
+	amino *Codec
+}
+
+// NewBaseCodec returns a reference to a new BaseCodec that implements the Marshaler
+// interface. A BaseCodec it used to support both Protocol Buffer and Amino
+// encoding. If no Amino codec is provided, it is assumed all encoding is done
+// via the type implementing the ProtoMarshaler interface (i.e. Protobuf).
+func NewBaseCodec(amino *Codec) Marshaler {
+	return &BaseCodec{amino}
+}
+
+// MarshalBinaryBare attempts to encode the provided type. If the type implements
+// ProtoMarshaler, serialization is delegated to Marshal. Otherwise, encoding falls
+// back on Amino.
+func (bc *BaseCodec) MarshalBinaryBare(o interface{}) ([]byte, error) {
+	m, ok := o.(ProtoMarshaler)
```

Comment:
So the reason I didn't take this approach (I did consider it) is because it is hard to tell at compile time what is actually going to happen. I mean you can consider every type and see whether it implements `ProtoMarshaler` or not, but I suspect this sort of indeterminism based on whether something implements an interface will be problematic. Which is why I took the approach of not trying to make the proto codec actually an interface that also accommodates amino.